activation/activation.go

```diff
@@ -403,26 +328,53 @@ func (b *Builder) buildNIPostChallenge(ctx context.Context) (*types.NIPostChalle
 		return nil, ctx.Err()
 	case <-b.syncer.RegisterForATXSynced():
 	}
+	current := b.currentEpoch()
+	prev, err := b.cdb.GetLastAtx(b.nodeID)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNotFound) {
+			return nil, err
+		}
+	} else if prev.PublishEpoch == current+1 {
+		current += 1
+	}
 
-	atxID, pubEpoch, err := b.GetPositioningAtxInfo()
+	until := time.Until(b.poetRoundStart(current))
+	if until > 0 {
+		metrics.PublishOntimeWindowLatency.Observe(until.Seconds())
+	}
+	if until > b.poetCfg.GracePeriod {
+		wait := until - b.poetCfg.GracePeriod
+		b.log.WithContext(ctx).With().Debug("waiting for fresh atxs",
+			log.Duration("till poet round", until),
+			log.Uint32("current epoch", current.Uint32()),
+			log.Duration("wait", wait),
+		)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(wait):
+		}
+	} else if until <= 0 {
+		metrics.PublishLateWindowLatency.Observe(-until.Seconds())
+		return nil, fmt.Errorf("%w: builder doesn't have time to submit in epoch %d. poet round already started %v ago",
+			ErrATXChallengeExpired, current, -until)
+	}
```

Comment:
I prefer checking for the error case first:

```suggestion
	until := time.Until(b.poetRoundStart(current))
	if until <= 0 {
		metrics.PublishLateWindowLatency.Observe(-until.Seconds())
		return nil, fmt.Errorf("%w: builder doesn't have time to submit in epoch %d. poet round already started %v ago",	ErrATXChallengeExpired, current, -until)
	}
	metrics.PublishOntimeWindowLatency.Observe(until.Seconds())
	if until > b.poetCfg.GracePeriod {
		wait := until - b.poetCfg.GracePeriod
		b.log.WithContext(ctx).With().Debug("waiting for fresh atxs",
			log.Duration("till poet round", until),
			log.Uint32("current epoch", current.Uint32()),
			log.Duration("wait", wait),
		)
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(wait):
		}
	}
```